metrics/controller/action.go

```diff
@@ -0,0 +1,159 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	dvoting "github.com/dedis/d-voting"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.dedis.ch/dela"
+	"go.dedis.ch/dela/cli/node"
+	"golang.org/x/xerrors"
+)
+
+// parameters to wait for the http server to start
+const (
+	defaultRetry = 10
+	waitRetry    = time.Second * 2
+)
+
+type startAction struct{}
+
+// Execute implements node.ActionTemplate. It registers the Prometheus handler.
+func (a startAction) Execute(ctx node.Context) error {
+	listenAddr := ctx.Flags.String("addr")
+	path := ctx.Flags.String("path")
+
+	fmt.Fprintf(ctx.Out, "starting metric server on %s %s", listenAddr, path)
+
+	// We'll go with the default Prometheus registerer
+	reg := prometheus.DefaultRegisterer
+
+	// Register Dela collectors
+	for _, c := range dela.PromCollectors {
+		err := reg.Register(c)
+		if err != nil {
+			fmt.Fprintf(ctx.Out, "ERROR: failed to register: %v", err)
+		}
+	}
+
+	// Register D-Voting collectors
+	for _, c := range dvoting.PromCollectors {
+		err := reg.Register(c)
+		if err != nil {
+			fmt.Fprintf(ctx.Out, "ERROR: failed to register: %v", err)
+		}
+	}
+
+	// initialize a new HTTP sever and register the Prometheus handler
+	srv := newMetricSrv(listenAddr)
+
+	srv.mux.Handle(path, promhttp.Handler())
+	fmt.Fprintf(ctx.Out, "registered prometheus service on %q", path)
+
+	// start the server and wait a bit to check for potential errors
+	errc := make(chan error, 1)
+
+	go func() {
+		errc <- srv.listen()
+	}()
+
+	for i := 0; i < defaultRetry && srv.getAddr() == nil; i++ {
+		select {
+		case err := <-errc:
+			return xerrors.Errorf("failed to listen: %v", err)
+		case <-time.After(waitRetry):
+		}
+	}
+
+	srvAddr := srv.getAddr()
+	if srvAddr == nil {
+		return xerrors.Errorf("failed to start metric server")
+	}
```

Comment:
According to [this](https://stackoverflow.com/a/32742904), the polling seems entirely avoidable.

Reply:
That doesn't solve our problem because we are using an external function to create and start the HTTP server. And we don't want the function to panic with a `log.Fatal(server.Serve(ln)) `.